### PR TITLE
GH-187 (partial PR): configurable routing - core mechanism

### LINKED ISSUE #187: Configurable Nested Routes and auxiliary router outlets
It's a subtask of #3.

After doing this task, all acceptance criteria of tasks mentioned above should hold. However, now we would like to also support configurable nested routes. Algorithms to find best matching url to render a link should be modified, if needed.

Motivation for having nested routes: Me may need nested routes to handle a 'sub-navigation' state of a page (for example the current step of the checkout or the active tab in product-details page).

It would be good to also investigate auxiliary router outlets

### PR DESCRIPTION
**Pull request by Platonn**
Close #187

### CODE REVIEW

File: projects/core/src/routing/configurable-routes/index.ts [hunk 1]
```diff
@@ -1,6 +1,8 @@
+export * from './config/configurable-routes-config';
+export * from './config/storefront-routes-translations';
+export * from './url-translator/index';
+
 export * from './configurable-routes.service';
 export * from './configurable-routes.module';
-export * from './configurable-routes-config';
 export * from './configurable-routes.service';
```
[the remark this answers is not in the document]
Comment (Platonn) [reply projects/core/src/routing/configurable-routes/index.ts]: Good catch!

File: projects/storefrontlib/src/lib/ui/layout/main/storefront.component.ts [hunk 2]
```diff
@@ -10,6 +10,6 @@ export class StorefrontComponent implements OnInit {
   constructor(private configurableRoutesService: ConfigurableRoutesService) {}
 
   ngOnInit() {
-    this.configurableRoutesService.changeLanguage('default'); // TODO #186: subscribe to active language from facade and pass it here
+    this.configurableRoutesService.translateRouterConfig('en');
```
[the remark this answers is not in the document]
Comment (Platonn) [reply projects/storefrontlib/src/lib/ui/layout/main/storefront.component.ts]: Indeed, we plan invoke translation of router config by APP_INITIALIZER in some next ticket.

File: projects/core/src/routing/configurable-routes/configurable-routes.service.ts [hunk 3]
```diff
@@ -12,79 +12,174 @@ export class ConfigurableRoutesService {
...
+
+  private translateRoute(
+    route: Route,
+    routesTranslations: RoutesTranslations
+  ): Routes {
+    if (this.isConfigurable(route, 'cxPath')) {
+      // we assume that 'path' and 'redirectTo' cannot be both configured for one route
+      if (
+        this.isConfigurable(route, 'cxRedirectTo') &&
+        !this.config.production
```
[the remark this answers is not in the document]
Comment (Platonn) [reply projects/core/src/routing/configurable-routes/configurable-routes.service.ts]: Thank you, I will remove it

File: projects/core/src/routing/configurable-routes/configurable-routes.service.ts [hunk 4]
```diff
@@ -95,28 +190,81 @@ export class ConfigurableRoutesService {
...
-        path: translatedPath
-      });
-    });
+  private translateRoutePath(
+    route: Route,
+    routesTranslations: RoutesTranslations
+  ): Route[] {
+    return this.getTranslatedPaths(route, 'cxPath', routesTranslations).map(
+      translatedPath => {
+        return Object.assign({}, route, {
```
[the remark this answers is not in the document]
Comment (Platonn) [reply projects/core/src/routing/configurable-routes/configurable-routes.service.ts]: Good point. I will change it.

File: projects/core/src/routing/configurable-routes/configurable-routes.service.ts [hunk 5]
```diff
@@ -95,28 +190,81 @@ export class ConfigurableRoutesService {
...
+    route: Route,
+    translations: RoutesTranslations
+  ): Route[] {
+    const translatedPaths = this.getTranslatedPaths(
+      route,
+      'cxRedirectTo',
+      translations
+    );
     return translatedPaths.length
       ? [Object.assign({}, route, { redirectTo: translatedPaths[0] })] // take the first path from list by convention
```
[the remark this answers is not in the document]
Comment (Platonn) [reply projects/core/src/routing/configurable-routes/configurable-routes.service.ts]: Good point. I will change it.

File: projects/core/src/routing/configurable-routes/url-translator/translate-url.pipe.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import { Pipe, PipeTransform } from '@angular/core';
+import { UrlTranslatorService } from './url-translator.service';
+
+@Pipe({
+  name: 'cxTranslateUrl'
+})
+export class TranslateUrlPipe implements PipeTransform {
```
[the remark this answers is not in the document]
Comment (Platonn) [reply projects/core/src/routing/configurable-routes/url-translator/translate-url.pipe.ts]: Sorry, I missed that. Thank you!

File: projects/core/src/routing/configurable-routes/url-translator/url-translator.service.ts [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    const [leafNestedRouteTranslation] = nestedRoutesTranslations.slice(-1);
+
+    // if no route was switched off in config (set to null), return root url:
+    if (!leafNestedRouteTranslation.paths) {
+      return this.ROOT_URL;
+    }
+
+    // find first path for every nested route that can staisfy it's parameters with given parameters
```
[the remark this answers is not in the document]
Comment (Platonn) [reply projects/core/src/routing/configurable-routes/url-translator/url-translator.service.ts]: You read carefully, thank you :)

File: projects/core/src/routing/configurable-routes/config/default-storefront-routes-translations.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  default?: StorefrontRoutesTranslations;
+  [languageCode: string]: StorefrontRoutesTranslations;
+} = {
+  default: {
+    homepage: { paths: [''] },
+    cart: { paths: ['cart'] },
+    search: { paths: ['search/:query'] },
+    login: { paths: ['login'] },
+    register: { paths: ['register'] },
+    resetNewPassword: { paths: ['reset-new-password/:token'] },
```
Comment (tobi-or-not-tobi) [inline projects/core/src/routing/configurable-routes/config/default-storefront-routes-translations.ts]: forgotPassword

File: projects/core/src/routing/configurable-routes/config/default-storefront-routes-translations.ts [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        allStores: { paths: ['view-all-stores'] },
+        listStores: {
+          paths: ['country/:country/region/:region', 'country/:country']
+        },
+        storeDescription: {
+          paths: ['country/:country/region/:region/:store']
+        }
+      }
+    },
+    termsAndConditions: { paths: ['terms-and-conditions'] },
```
Comment (tobi-or-not-tobi) [inline projects/core/src/routing/configurable-routes/config/default-storefront-routes-translations.ts]: as discussed, the following are CMS driven and should not be part of the hard-coded config:
- termsAndConditions
- contact
- help
- sale

File: projects/core/src/routing/configurable-routes/config/default-storefront-routes-translations.ts (hunk 8 above)
Comment (Platonn) [reply projects/core/src/routing/configurable-routes/config/default-storefront-routes-translations.ts]: I renamed config key names: `resetPassword` to `forgotPassword` nad `resetNewPassword` to `resetPassword`.
However, I left paths untouched.

File: projects/core/src/routing/configurable-routes/config/default-storefront-routes-translations.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
+import { StorefrontRoutesTranslations } from './storefront-routes-translations';
+
+export const defaultStorefrontRoutesTranslations: {
+  default?: StorefrontRoutesTranslations;
+  [languageCode: string]: StorefrontRoutesTranslations;
+} = {
+  default: {
+    homepage: { paths: [''] },
```
[the remark this answers is not in the document]
Comment (Platonn) [reply projects/core/src/routing/configurable-routes/config/default-storefront-routes-translations.ts]: Good point, thank you

File: projects/core/src/routing/configurable-routes/config/default-storefront-routes-translations.ts (hunk 9 above)
Comment (Platonn) [reply projects/core/src/routing/configurable-routes/config/default-storefront-routes-translations.ts]: I'll put a note in documentation.
And as discussed offline, I removed `myAccount_` prefix from config keys

**Comment by Platonn** [comment]
Thanks for review. I close this PR without merge. This code will be merged in one big PR #680

**Pull request closed without merging**